daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala

```diff
@@ -30,66 +25,51 @@ class GenDarReader[A](
       name: String,
       darStream: ZipInputStream,
       entrySizeThreshold: Int = EntrySizeThreshold,
-  ): Try[Dar[A]] = {
+  ): Try[Dar[A]] =
     for {
-      entries <- bracket(Try(darStream))(zis => Try(zis.close())).flatMap(zis =>
-        loadZipEntries(name, zis, entrySizeThreshold)
-      )
-      names <- entries.readDalfNames(DarManifestReader.dalfNames): Try[Dar[String]]
-      main <- parseOne(entries.getInputStreamFor)(names.main): Try[A]
-      deps <- parseAll(entries.getInputStreamFor)(names.dependencies): Try[List[A]]
+      entries <- loadZipEntries(name, darStream, entrySizeThreshold)
+      manifestContent <- entries.get(ManifestName)
+      names <- bracket(Try(manifestContent.toInputStream))(is => Try(is.close()))
+        .flatMap(DarManifestReader.dalfNames)
+      main <- parseOne(entries.get)(names.main)
+      deps <- parseAll(entries.get)(names.dependencies)
     } yield Dar(main, deps)
-  }
 
   // Fails if a zip bomb is detected
+  @throws[Error.ZipBomb]
+  @throws[IOException]
   private def slurpWithCaution(
+      name: String,
       zip: ZipInputStream,
       entrySizeThreshold: Int,
-  ): Try[(Long, InputStream)] =
-    Try {
-      val output = new ByteArrayOutputStream()
-      val buffer = new Array[Byte](4096)
-      for (n <- Iterator.continually(zip.read(buffer)).takeWhile(_ >= 0) if n > 0) {
-        output.write(buffer, 0, n)
-        if (output.size >= entrySizeThreshold) throw Error.ZipBomb()
-      }
-      (output.size.toLong, new ByteArrayInputStream(output.toByteArray))
-    }
+  ): (String, Bytes) = {
+    val output = Bytes.fromInputStream(zip, entrySizeThreshold)
+    if (zip.read() >= 0) throw Error.ZipBomb()
+    name -> output
+  }
 
   private def loadZipEntries(
       name: String,
       darStream: ZipInputStream,
       entrySizeThreshold: Int,
-  ): Try[ZipEntries] = {
-    @tailrec
-    def go(accT: Try[Map[String, (Long, InputStream)]]): Try[Map[String, (Long, InputStream)]] =
-      Option(darStream.getNextEntry) match {
-        case Some(entry) =>
-          go(
-            accT.flatMap { acc =>
-              bracket(slurpWithCaution(darStream, entrySizeThreshold))(_ =>
-                Try(darStream.closeEntry())
-              )
-                .map { sizedBytes =>
-                  acc + (entry.getName -> sizedBytes)
-                }
-            }
-          )
-        case None => accT
-      }
-
-    go(Success(Map.empty)).map(ZipEntries(name, _))
-  }
+  ): Try[ZipEntries] =
+    bracket(Try(darStream))(is => Try(is.close()))
```

Comment:
I’m not sure this is a good idea. imho a function should not close a stream it didn’t open and we didn’t open the zip stream here.

Reply:
967146aec86b70968347210b47261bd3c5006779